core/solver/multigrid.cpp

```diff
@@ -0,0 +1,595 @@
+
+
+#include <complex>
+#include <iostream>
+#include <typeinfo>
+
+
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/name_demangling.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/solver/ir.hpp>
+#include <ginkgo/core/stop/iteration.hpp>
+
+
+#include "core/components/fill_array.hpp"
+#include "core/solver/ir_kernels.hpp"
+#include "core/solver/multigrid_kernels.hpp"
+
+
+namespace gko {
+namespace solver {
+namespace multigrid {
+
+
+GKO_REGISTER_OPERATION(initialize, ir::initialize);
+GKO_REGISTER_OPERATION(fill_array, components::fill_array);
+GKO_REGISTER_OPERATION(kcycle_step_1, multigrid::kcycle_step_1);
+GKO_REGISTER_OPERATION(kcycle_step_2, multigrid::kcycle_step_2);
+GKO_REGISTER_OPERATION(kcycle_check_stop, multigrid::kcycle_check_stop);
+
+
+}  // namespace multigrid
+
+
+namespace {
+
+template <template <typename> class Base, typename T, typename func,
+          typename... Args>
+void run(T obj, func, Args... args)
+{
+    GKO_NOT_IMPLEMENTED;
+}
+
+
+template <template <typename> class Base, typename K, typename... Types,
+          typename T, typename func, typename... Args>
+void run(T obj, func f, Args... args)
+{
+    if (auto dobj = std::dynamic_pointer_cast<const Base<K>>(obj)) {
+        f(dobj, args...);
+    } else {
+        run<Base, Types...>(obj, f, args...);
+    }
+}
+
+template <typename ValueType, typename T>
+std::enable_if_t<is_complex_s<ValueType>::value == is_complex_s<T>::value,
+                 ValueType>
+casting(const T &x)
+{
+    return static_cast<ValueType>(x);
+}
+
+template <typename ValueType, typename T>
+std::enable_if_t<!is_complex_s<ValueType>::value && is_complex_s<T>::value,
+                 ValueType>
+casting(const T &x)
+{
+    return static_cast<ValueType>(real(x));
+}
+
+template <typename ValueType>
+void handle_list(
+    size_type index, std::shared_ptr<const LinOp> &matrix,
+    std::vector<std::shared_ptr<const LinOpFactory>> &smoother_list,
+    std::vector<std::shared_ptr<LinOp>> &smoother, size_type iteration,
+    std::complex<double> relaxation_factor)
+{
+    auto list_size = smoother_list.size();
+    if (list_size != 0) {
+        auto temp_index = list_size == 1 ? 0 : index;
+        GKO_ENSURE_IN_BOUNDS(temp_index, list_size);
+        auto item = smoother_list.at(temp_index);
+        if (item == nullptr) {
+            smoother.emplace_back(nullptr);
+        } else {
+            auto solver = item->generate(matrix);
+            if (solver->apply_uses_initial_guess() == true) {
+                smoother.emplace_back(give(solver));
+            } else {
+                auto ir = smoother_build<ValueType>(
+                    give(solver), iteration,
+                    casting<ValueType>(relaxation_factor));
```

Comment:
Do you think it might make sense to have different relaxation factors for different smoother levels ? I think to accelerate convergence it might be useful to have more relaxed coarse levels.

Reply:
Yes, but it can be handled by smoother_list